### ISSUE TITLE
--interactive=false does not work on v2.4.1

### ISSUE THREAD

**Issue by nightpool** [docker/compose#9421]
**Description**

I'm trying to run a container with `in: :close` from Ruby, so that I can automate a handful of commonly used docker compose commands (like build, up, and run) in parallel with each other, without them confusingly sharing a standard input. This seems to be a regression since #9246, or at least an incomplete implementation of the stdinopen feature. This is a regression / undocumented breaking change from docker-compose v1

**Steps to reproduce the issue:**
Create a new docker-compose with an "app" service.

Run the following ruby code:
```ruby
command = 'docker-compose run --no-deps --rm -T --interactive=false app echo 1'.split(" ")
system(*command, in: :close) # Create docker-compose subprocess with stdin closed, fails
system(*command) # It works when stdin is open
```

**Describe the results you received:**

The first subprocess call fails with "fork/exec /usr/local/bin/com.docker.cli: bad file descriptor". The second call succeeds.

**Describe the results you expected:**

Both calls succeed, and I can create `docker-compose` subprocesses both with and without stdin attached.

**Additional information you deem important (e.g. issue happens only occasionally):**

This worked on docker-compose v1 and is a blocker for upgrading from V1 to V2. 

**Output of `docker compose version`:**

```
Docker Compose version v2.4.1
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.8.2)
  compose: Docker Compose (Docker Inc., v2.4.1)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 13
  Running: 9
  Paused: 0
  Stopped: 4
 Images: 28
 Server Version: 20.10.14
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3df54a852345ae127d1fa3092b95168e4a88e2f8
 runc version: v1.0.3-0-gf46b6ba
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.104-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.64GiB
 Name: docker-desktop
 ID: [redacted]
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 125
  Goroutines: 112
  System Time: 2022-04-27T20:42:53.818214331Z
  EventsListeners: 4
```

**Additional environment details:**

I'm on Docker for Mac, and i've enabled the new virtualization framework options as well as the virtiofs options. Neither seem to have an impact on this specific issue though, as far as I can tell.

**Comment by nightpool** [comment]
@ndeloof I know you implemented the `interactive=false` feature... is this on your radar? It seems like it kind of defeats the purpose of adding the `--interactive=false` command if you still need to have stdin open to run the container, but not sure if that use-case was on your radar.

**Comment by x-yuri** [comment]
Also, it's not in the `exec`'s usage:

```
$ docker compose exec --help

Usage:  docker compose exec [options] [-e KEY=VAL...] [--] SERVICE COMMAND [ARGS...]

Execute a command in a running container.

Options:
  -d, --detach                       Detached mode: Run command in the background.
  -e, --env stringArray              Set environment variables
      --index int                    index of the container if there are multiple instances of a service [default: 1].
                                     (default 1)
  -T, --no-TTY docker compose exec   Disable pseudo-TTY allocation. By default docker compose exec allocates a TTY.
                                     (default true)
      --privileged                   Give extended privileges to the process.
  -u, --user string                  Run the command as this user.
  -w, --workdir string               Path to workdir directory for this command.
```

Initially I did:

```
print_fnames | while IFS= read -r fname; do
    docker compose exec -T C touch "$fname"
done
```

But `docker-compose` read part of the lines (filenames), so I did:

```
print_fnames | while IFS= read -r fname; do
    docker compose exec -T C touch "$fname" </dev/null
done
```

With `--interactive=false` it also works:

```
print_fnames | while IFS= read -r fname; do
    docker compose exec -T --interactive=false C touch "$fname"
done
```

Maybe `</dev/null` will make it work in your case too?..

**Comment by ndeloof** [comment]
`--interactive` and `--tty` where introduced as we noticed many users did use with compose the same `exec -it ...` command they use with `docker` command line. As this is the default behavior, those were set for convenience and hidden from help.

@nightpool it seems to me this issue is caused by the wrapper set by Docker Desktop around `com.docker.cli` (the actual docker CLI).
Could you try doing the same running `com.docker.cli   compose`  as a replacement for `docker-compose` command ?

**Comment by ndeloof** [comment]
I ran a quick test, and was able to get it work when invoking directly docker-compose "CLI plugin" as a standalone binary:
```ruby
command = '/Users/nicolas/.docker/cli-plugins/docker-compose run -T --interactive=false test echo 1'.split(" ")
system(*command, in: :close) 
```

```
$ ruby toto.rb
1
```


Works also well if I build Docker CLI from sources:

```ruby
command = '/Users/nicolas/go/src/github.com/docker/cli/build/docker-darwin-arm64 compose run -T --interactive=false test echo 1'.split(" ")
system(*command, in: :close) 
```

```
$ ruby toto.rb
1
```

but fails when I use Docker Desktop docker CLI:
```ruby
command = '/Applications/Docker.app/Contents/Resources/bin/com.docker.cli compose run -T --interactive=false test echo 1'.split(" ")
system(*command, in: :close) 
```

```
$ ruby toto.rb
fork/exec /Users/nicolas/.docker/cli-plugins/docker-compose: bad file descriptor
```

**Comment by ndeloof** [comment]
confirmed this is fixed when using docker CLI 23.0.0-beta.1 (23.x will be released soon-ish)
Haven't found which exact commit on docker CLI allows this, feel free to play with git bisect :P

**Comment by nightpool** [comment]
Thanks for the update!
